proposals/NNNN-atomics.md

```diff
@@ -0,0 +1,1702 @@
+    ordering: AtomicUpdateOrdering
+  ) -> (exchanged: Bool, original: Value)
+
+  public func compareExchange(
+    expected: Value,
+    desired: __owned Value,
+    ordering: AtomicUpdateOrdering,
+    failureOrdering: AtomicLoadOrdering
+  ) -> (exchanged: Bool, original: Value)
+
+  public func weakCompareExchange(
+    expected: Value,
+    desired: __owned Value,
+    ordering: AtomicUpdateOrdering,
+    failureOrdering: AtomicLoadOrdering
+  ) -> (exchanged: Bool, original: Value)
+}
+```
+
+The `ordering` arguments indicate if the atomic operation is also expected to synchronize the effects of previous (or subsequent) accesses. This is explained in a separate section below.
+
+The first three operations are relatively simple:
+
+- `load` returns the current value.
+- `store` updates it.
+- `exchange` is a combination of `load` and `store`; it updates the
+  current value and returns the previous one as a single atomic
+  operation.
+
+The three `compareExchange` variants are somewhat more complicated: they implement a version of `exchange` that only performs the update if the original value is the same as a supplied expected value. To be specific, they execute the following algorithm as a single atomic transaction:
+
+```swift
+  guard currentValue == expected else { 
+    return (exchanged: false, original: currentValue) 
+  }
+  currentValue = desired
+  return (exchanged: true, original: expected)
+```
+
+All three variants implement the same algorithm. The first variant uses the same memory ordering whether or not the exchange succeeds, while the other two allow callers to specify two distinct memory orderings for the success and failure cases. The two orderings are independent from each other -- all combinations of update/load orderings are supported [[P0418]]. (Of course, the implementation may need to "round up" to the nearest ordering combination that is supported by the underlying code generation layer and the targeted CPU architecture.)
+
+The `weakCompareExchange` form may sometimes return false even when the original and expected values are equal. (Such failures may happen when some transient condition prevents the underlying operation from succeeding -- such as an incoming interrupt during a load-link/store-conditional instruction sequence.) This variant is designed to be called in a loop that only exits when the exchange is successful; in such loops using `weakCompareExchange` may lead to a performance improvement by eliminating a nested loop in the regular, "strong", `compareExchange` variants.
+
+The compare-exchange primitive is special: it is a universal operation that can be used to implement all other atomic operations, and more. For example, here is how we could use `compareExchange` to implement a wrapping increment operation over `UnsafeAtomic<Int>` values:
+
+```swift
+extension UnsafeAtomic where Value == Int {
+  func wrappingIncrement(
+    by operand: Int,
+    ordering: AtomicUpdateOrdering
+  ) -> Int {
```

Comment:
No return value?

```suggestion
  ) {
```

Reply:
This is quite intentional. There are separate variants for increments/decrements with a return value.

Reply:
So, should the `-> Int` be removed for this variant?

Reply:
🤦‍♂️